### PR TITLE
Finish Firestore equality implementations

### LINKED ISSUE #1883: Firestore: Implement equality for Firestore public types
Precise semantics are being discussed in the Firestore team.

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #1883.

One type per commit for ease of reviewing.

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs [hunk 1]
```diff
@@ -65,5 +66,19 @@ public sealed class Precondition
...
+        /// Compares this precondition with another for equality.
+        /// </summary>
+        /// <param name="other">The precondition to compare this one with</param>
+        /// <returns><c>true</c> if this precondition is equal to <paramref name="other"/>; <c>false</c> otherwise.</returns>
+        public bool Equals(Precondition other) => other != null && Equals(Proto, other.Proto);
+
+        /// <inheritdoc />
+        public override bool Equals(object obj) => Equals(obj as Precondition);
+
+        // Value for null is arbitrary, but not 0 as that's what a MustExist gives too.
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: nit: Can this comment be explained better? I don't understand what this means.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: Sure, will do. It's explaining the "?? 5172" bit, but I can make it clearer.

Comment (chrisdunelm) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: I got the `?? 5172` bit; the "that's what MustExist gives too" bit confused me.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: Ah, I see. It so happens that MustExist.Proto.GetHashCode() returns 0, so it's best if None (no proto) returns something other than 0.

**Pull request merged**